File: library/src/main/java/com/owncloud/android/lib/resources/e2ee/UnlockFileRemoteOperation.java
/*
 * Nextcloud Android Library
 *
 * SPDX-FileCopyrightText: 2017-2024 Nextcloud GmbH and Nextcloud contributors
 * SPDX-FileCopyrightText: 2017 Tobias Kaminsky <tobias@kaminsky.me>
 * SPDX-License-Identifier: MIT
 */
package com.owncloud.android.lib.resources.e2ee;

import com.owncloud.android.lib.common.OwnCloudClient;
import com.owncloud.android.lib.common.operations.RemoteOperation;
import com.owncloud.android.lib.common.operations.RemoteOperationResult;
import com.owncloud.android.lib.common.utils.Log_OC;

import org.apache.commons.httpclient.HttpStatus;
import org.apache.commons.httpclient.methods.DeleteMethod;


/**
 * Unlock a file
 */
public class UnlockFileRemoteOperation extends RemoteOperation<Void> {

    private static final String TAG = UnlockFileRemoteOperation.class.getSimpleName();
    private static final int SYNC_READ_TIMEOUT = 40000;
    private static final int SYNC_CONNECTION_TIMEOUT = 5000;
    private static final String LOCK_FILE_URL = "/ocs/v2.php/apps/end_to_end_encryption/api/v2/lock/";

    private final long localId;
    private final String token;

    /**
     * Constructor
     */
    public UnlockFileRemoteOperation(long localId, String token) {
        this.localId = localId;
        this.token = token;
    }

    /**
     * @param client Client object
     */
    @Override
    protected RemoteOperationResult<Void> run(OwnCloudClient client) {
        RemoteOperationResult<Void> result;
        DeleteMethod deleteMethod = null;

        try {
            // remote request
            deleteMethod = new DeleteMethod(client.getBaseUri() + LOCK_FILE_URL + localId);
            deleteMethod.addRequestHeader(OCS_API_HEADER, OCS_API_HEADER_VALUE);
            deleteMethod.addRequestHeader(CONTENT_TYPE, FORM_URLENCODED);
            deleteMethod.addRequestHeader(E2E_TOKEN, token);

            int status = client.executeMethod(deleteMethod, SYNC_READ_TIMEOUT, SYNC_CONNECTION_TIMEOUT);

            result = new RemoteOperationResult<>(status == HttpStatus.SC_OK, deleteMethod);
            
            client.exhaustResponse(deleteMethod.getResponseBodyAsStream());
        } catch (Exception e) {
            result = new RemoteOperationResult<>(e);
            Log_OC.e(TAG, "Unlock file with id " + localId + " failed: " + result.getLogMessage(),
                     result.getException());
        } finally {
            if (deleteMethod != null)
                deleteMethod.releaseConnection();
        }
        return result;
    }
}

File: library/src/main/java/com/owncloud/android/lib/resources/e2ee/UnlockFileRemoteOperation.kt
/*
 * Nextcloud Android Library
 *
 * SPDX-FileCopyrightText: 2026 Alper Ozturk <alper.ozturk@nextcloud.com>
 * SPDX-License-Identifier: MIT
 */
package com.owncloud.android.lib.resources.e2ee

import com.nextcloud.common.SessionTimeOut
import com.nextcloud.common.defaultSessionTimeOut
import com.owncloud.android.lib.common.OwnCloudClient
import com.owncloud.android.lib.common.operations.RemoteOperation
import com.owncloud.android.lib.common.operations.RemoteOperationResult
import com.owncloud.android.lib.common.utils.Log_OC
import org.apache.commons.httpclient.HttpStatus
import org.apache.commons.httpclient.methods.DeleteMethod

class UnlockFileRemoteOperation
    @JvmOverloads
    constructor(
        private val localId: Long,
        private val token: String,
        private val sessionTimeOut: SessionTimeOut = defaultSessionTimeOut,
        private val useV2: Boolean = true
    ) : RemoteOperation<Void>() {
        @Deprecated("Deprecated in Java")
        @Suppress("Detekt.TooGenericExceptionCaught")
        override fun run(client: OwnCloudClient): RemoteOperationResult<Void> =
            runCatching { unlockFile(client) }
                .getOrElse { e ->
                    Log_OC.e(TAG, "Unlock file with id $localId failed", e)
                    RemoteOperationResult(e as? Exception ?: RuntimeException(e))
                }

        private fun unlockFile(client: OwnCloudClient): RemoteOperationResult<Void> {
            val (status, deleteMethod) = executeWithFallback(client)

            return RemoteOperationResult<Void>(status == HttpStatus.SC_OK, deleteMethod).also {
                client.exhaustResponse(deleteMethod.responseBodyAsStream)
                deleteMethod.releaseConnection()
            }
        }

        /**
         * Tries the v2 endpoint first, falling back to v1 on 404/500.
         * Returns the final status code and the method used.
         */
        private fun executeWithFallback(client: OwnCloudClient): Pair<Int, DeleteMethod> {
            val v2Method = buildDeleteMethod(client, LOCK_FILE_URL_V2)
            val v2Status = client.executeMethod(v2Method, sessionTimeOut.readTimeOut, sessionTimeOut.connectionTimeOut)

            val needsFallback =
                (!useV2 || v2Status == HttpStatus.SC_NOT_FOUND || v2Status == HttpStatus.SC_INTERNAL_SERVER_ERROR)
            if (!needsFallback) return v2Status to v2Method

            v2Method.releaseConnection()
            val v1Method = buildDeleteMethod(client, LOCK_FILE_URL_V1)
            val v1Status = client.executeMethod(v1Method, sessionTimeOut.readTimeOut, sessionTimeOut.connectionTimeOut)
            return v1Status to v1Method
        }

        private fun buildDeleteMethod(
            client: OwnCloudClient,
            baseUrl: String
        ) = DeleteMethod("${client.baseUri}$baseUrl$localId").apply {
            addRequestHeader(OCS_API_HEADER, OCS_API_HEADER_VALUE)
            addRequestHeader(CONTENT_TYPE, FORM_URLENCODED)
            addRequestHeader(E2E_TOKEN, token)
        }

        companion object {
            private val TAG = UnlockFileRemoteOperation::class.java.simpleName
            private const val LOCK_FILE_URL_V1 = "/ocs/v2.php/apps/end_to_end_encryption/api/v1/lock/"
            private const val LOCK_FILE_URL_V2 = "/ocs/v2.php/apps/end_to_end_encryption/api/v2/lock/"
        }
    }
